example_envs/bookinfo/gateway.rs

```diff
@@ -20,22 +25,54 @@ impl fmt::Display for Gateway {
 
 impl SimElement for Gateway {
     fn tick(&mut self, tick: u64) -> Vec<Rpc> {
-        let mut outbound_rpcs = self.core_node.tick(tick);
-        for outbound_rpc in &mut outbound_rpcs {
-            outbound_rpc
-                .headers
-                .insert("direction".to_string(), "request".to_string());
-            outbound_rpc
-                .headers
-                .insert("src".to_string(), self.core_node.id.to_string());
-            outbound_rpc
-                .headers
-                .insert("dest".to_string(), "productpage-v1".to_string());
+        log::error!("{:?}", self.core_node.generation_rate);
+        for _ in 0..max(
+            self.core_node.ingress_queue.size(),
+            self.core_node.generation_rate as usize,
+        ) {
+            let mut queued_rpcs: Vec<Rpc> = vec![];
+            // Dequeue an RPC, or generate one
+            let mut rpc: Rpc;
+            if let Some(deq) = self.core_node.dequeue_ingress(tick) {
+                rpc = deq;
+            } else {
+                rpc = Rpc::new(&tick.to_string());
+                rpc.headers
+                    .insert("direction".to_string(), "request".to_string());
+            }
+
+            // Select the destination
+            let mut new_rpcs: Vec<Rpc> = vec![];
+            self.process_rpc(&mut rpc, &mut new_rpcs);
+
+            // Pass the RPCs we have through the plugin
+            for rpc in new_rpcs {
+                self.core_node
+                    .pass_through_plugin(rpc, &mut queued_rpcs, tick, "egress");
+            }
+            for queued_rpcs in &queued_rpcs {
+                self.core_node.enqueue_egress(queued_rpcs.clone())
+            }
+        }
+        let max_output = min(
+            self.core_node.egress_queue.size(),
+            self.core_node.egress_rate as usize,
+        );
+        let mut outbound_rpcs: Vec<Rpc> = vec![];
+        for _ in 0..max_output {
+            outbound_rpcs.push(self.core_node.dequeue_egress().unwrap())
         }
         outbound_rpcs
     }
-    fn recv(&mut self, _rpc: Rpc, _tick: u64) {
-        // we discard anything we receive
+    fn recv(&mut self, rpc: Rpc, tick: u64) {
+        if let Some(direction) = rpc.headers.get("direction") {
+            if direction == "response" {
+                self.collected_responses.push(rpc);
+            }
+        } else {
+            // RPC without direction information, just pass it on
+            self.core_node.recv(rpc, tick);
```

Comment:
Shouldn't we raise an error if an RPC has no direction information?

Reply:
I do not know, gateway is kind of dumb so it fills that information anyway. If it gets a directionless RPC that is being forwarded to productpage.

Reply:
Okay, I guess the tests will guarantee that never happens anyway.